torch/distributed/fsdp/flat_param.py

```diff
@@ -1634,8 +1667,26 @@ def reshard(self, free_unsharded_flat_param: bool):
         # `use_orig_params=True`, the `param` does not point to valid memory
         # when setting `param.data = ...` in `_use_sharded_views()`.
         self._use_sharded_flat_param()
+
         if free_unsharded_flat_param:
-            self._free_unsharded_flat_param()
+            if self._limit_all_gathers:
+                # In the new rate limiter, we stash an all-gather buffer to keep
+                # it from being freed by the caching allocator until two
+                # all-gathers later. So that the caching allocator would try to
+                # reuse its storage for the +2 all-gather.
+                # We also do not call `record_stream` to hand over the ownership
+                # to the compute stream (i.e. the ownership is still with the
+                # unshard stream). We keep the tensor alive via stashing.
+                comp_done_event = self._device_handle.Event()
+                comp_done_event.record()
+                unsharded_flat_param = self._get_padded_unsharded_flat_param()
```

Comment:
Note that any Tensor that is a view into the storage will prevent the de-allocation (even if it looks at only a subset of the storage). 
So here you can take whichever Tensor that has a view into that storage so we should use whichever is cheapest to get.
The "padded" in the name here might mean that this is an expensive one to get?

Reply:
Thanks for sharing the knowledge!

`_get_padded_unsharded_flat_param` is called to get a reference to the actual unshard buffer used -- FSDP creates two such buffers: a "regular" one and a "full-precision" one. I guess the additional one was created to support mixed precision. At runtime, only one of these two unshard buffers is used based on some knobs. I guess the mixed precision support could have been done in a different way, but figured better not to touch it in this PR.